### PR TITLE
Add 'cyclop' linter 

### PR DESCRIPTION
**Pull request by bkielbasa**
https://github.com/bkielbasa/cyclop

### CODE REVIEW

**Comment by ldez** [comment]
Hello,

to be able to sign the CLA you have to add the email that you are using inside your GitHub account settings https://github.com/settings/emails

**Comment by ldez** [comment]
Could you explain what is for you the difference between your linter the other linters on the same subjects (like gocyclo, funlen, ...)?

Otherwise, could you run a `go mod tidy`

**Comment by bkielbasa** [comment]
Hi @ldez!

Of course :) Firstly, `gocognit` calculates cognitive complexity which is something a bit different than cyclomatic complexity. 
Secondly, the `gocyclo` is OK but the author seems to not maintain the project a lot. There were pull requests or issues waiting months to merge/review/answer. This is why I created my own. What's more, it's written in the "old fashion way" without the `go/analizer` package usage.
The `funlen` checks the length of the function which's something different than the cyclomatic complexity.

I hope I answered your questions! If you have more, don't hesitate to ask :)

**Comment by ldez** [comment]
> Secondly, the gocyclo is OK but the author seems to not maintain the project a lot. There were pull requests or issues waiting months to merge/review/answer. 

https://github.com/fzipp/gocyclo seems maintained, at least there is a lot of commits in the past year. https://github.com/fzipp/gocyclo/commits/main

File: pkg/lint/lintersdb/manager.go [hunk 1]
```diff
@@ -194,6 +194,9 @@ func (m Manager) GetAllSupportedLinterConfigs() []*linter.Config {
 		linter.NewConfig(golinters.NewGocyclo()).
 			WithPresets(linter.PresetComplexity).
 			WithURL("https://github.com/alecthomas/gocyclo"),
+		linter.NewConfig(golinters.NewCyclop()).
+			WithPresets(linter.PresetComplexity).
+			WithURL("https://github.com/bkielbasa/cyclop"),
```
Comment (ldez) [inline pkg/lint/lintersdb/manager.go:199]:
```suggestion
		linter.NewConfig(golinters.NewCyclop()).
			WithLoadForGoAnalysis().
			WithPresets(linter.PresetComplexity).
			WithURL("https://github.com/bkielbasa/cyclop"),
```

File: test/testdata/cyclop.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
+// args: -Ecyclop
+package testdata
+
+func cyclopComplexFunc() { // ERROR "calculated cyclomatic complexity for function cyclopComplexFunc is 11, max is 10"
+	i := 1
```
Comment (ldez) [inline test/testdata/cyclop.go:5]:
```suggestion
package testdata

import "math"

func cyclopComplexFunc() { // ERROR "calculated cyclomatic complexity for function cyclopComplexFunc is 11, max is 10"
	i := math.MaxInt8
```

File: pkg/golinters/cyclop.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+const cyclopName = "cyclop"
+
+func NewCyclop() *goanalysis.Linter {
+	return goanalysis.NewLinter(
+		cyclopName,
+		"checks function and package cyclomatic complexity",
+		[]*analysis.Analyzer{
+			analyzer.NewAnalyzer(),
+		},
```
Comment (ldez) [inline pkg/golinters/cyclop.go:18]:
```suggestion
		[]*analysis.Analyzer{analyzer.NewAnalyzer()},
```

**Review by ldez — CHANGES_REQUESTED** [review]

File: test/testdata/cyclop.go (hunk 2 above)
Comment (bombsimon) [reply test/testdata/cyclop.go:5]: What's the benefit with using `MaxInt8` and pulling in `math` as opposed to just using 1?

File: pkg/golinters/cyclop.go (hunk 3 above)
Comment (bombsimon) [reply pkg/golinters/cyclop.go:18]: Quite opinionated suggestion, do we have a standard for this? Seems like the authors pattern is used a lot, f.ex. in [gomnd](https://github.com/golangci/golangci-lint/blob/master/pkg/golinters/gomnd.go#L12-L14), [nlreturn](https://github.com/golangci/golangci-lint/blob/master/pkg/golinters/nlreturn.go#L14-L16), [tparallel](https://github.com/golangci/golangci-lint/blob/master/pkg/golinters/tparallel.go) and a bunch of other places in the source code. Not sure we should request changes for these kind of things?

File: test/testdata/cyclop.go (hunk 2 above)
Comment (ldez) [reply test/testdata/cyclop.go:5]: It's related to https://github.com/golangci/golangci-lint/pull/1733

**Comment by ldez** [comment]
@bkielbasa could you rebase your PR?

File: test/testdata/cyclop.go (hunk 2 above)
Comment (bombsimon) [reply test/testdata/cyclop.go]: Ah, right! Remember reading that now that you point it out - thanks! 🎉

**Comment by bkielbasa** [comment]
@ldez done!

**Review by ldez — APPROVED** [review]
LGTM

File: pkg/config/config.go [hunk 4]
```diff
@@ -183,6 +183,11 @@ type LintersSettings struct {
 	Gocyclo struct {
 		MinComplexity int `mapstructure:"min-complexity"`
 	}
+	Cyclop struct {
```
Comment (maratori) [inline pkg/config/config.go:186]: @bkielbasa Am I right that this config is never used?

Comment (ldez) [reply pkg/config/config.go:186]: @maratori https://github.com/golangci/golangci-lint/pull/1743

Comment (maratori) [reply pkg/config/config.go:186]: 👌

**Pull request merged**